src/06_ffi.md

```diff
@@ -0,0 +1,395 @@
+# Interfacing (FFI)
+
+Rust approach to interfacing with other programs relies on a strong
+compatibility with C. However, this boundary is by its very nature `unsafe`.
+
+Functions that are marked `extern` are made compatible with C code at the
+compilation. They may be called from C code with any parameter.
+The exact syntax is `extern "<ABI>"` where ABI is a calling convention and
+depends on the target platform. The default one is `C` which corresponds to
+a standard C calling convention on the target platform.
+
+```rust
+// export a C-compatible function
+#[no_mangle]
+pub extern "C" fn mylib_f(param: u32) -> i32 {
+    if param == 0xCAFEBABE { 0 } else { -1 }
+}
+```
+
+For the function `mylib_f` to be accessible with the same name, the function
+must also be annotated with the `#[no_mangle]` attribute.
+
+Conversely, one can call C functions from Rust if they are declared in an
+`extern` block:
+
+```rust
+// import an external function from libc
+extern "C" {
+    fn abs(args: i32) -> i32;
```

Comment:
This should be using `c_int`

Reply:
Which, on its own, also leads to another section, about, when not using `bindgen` for the `extern` declarations, to use C types such as those defined in `::libc`.

With a special mention against using uninhabited types (_e.g._, `enum Void {}`) to represent C's `void`.

Reply:
Thanks! Fix pushed for the example.

That's a good point! Something should be written about that.

Side note: ANSSI recommends to actually use fixed-size integer types in C (uint32_t, int64_t, etc.) when possible. They are directly compatible with Rust integer types (u32, i64, etc.).

Reply:
Indeed, there is nothing worse than C's `int` ... except for the fact that it is (sadly) quite pervasive in many C APIs. So the rule should be _use `c_int` when APIs expose / require\* `int`, but otherwise use fixed-size integers_.

\* _e.g._, for callbacks or "plugins"